Refactor VMChangeSet to abstract size for gas charging

igor-aptos opened the pull request:
### Description

- consolidate all the different new fields we were adding in VMChangeSet to support resource group / aggregator (delayed fields) gas charging
- simplify gas charging a bit
- make gas charging be based on size provided in the VMChangeSet (that can be different from the value inside, but needs to be the end materialized size)

- last commit in the stack makes AsSum resource group gas charging use the same formula that is used with flag disabled - making it so that replacing one variant of AbstractResourceWriteOp with a materialized version always keeps the size identical.

gelash commented on aptos-move/aptos-vm-types/src/storage.rs:
```diff
@@ -69,14 +69,14 @@ impl StoragePricingV1 {
 
         match op {
-            Creation(data) | CreationWithMetadata { data, .. } => {
+            Creation(value_size) => {
```
Would it make sense to match or map on the op.write_len() here? (since that's what determines the logic at the moment).

igor-aptos replied:
here at least the logic is different whetehr it is creation (additional write_data_per_new_item) or not

gelash commented on aptos-move/aptos-vm-types/src/storage.rs:
```diff
@@ -147,19 +147,19 @@ impl StoragePricingV2 {
 
         match &op {
-            Creation(data) | CreationWithMetadata { data, .. } => {
+            Creation(value_size) => {
```
Ditto: re: matching op.write_len()

igor-aptos replied:
it's per_byte_create vs per_byte_write

gelash commented on types/src/write_set.rs:
```diff
@@ -144,6 +144,40 @@ impl WriteOp {
     }
 }
 
+pub enum WriteOpSize {
```
Just a thought to consider having struct WriteOpSize with named deposit: Option<u64>, len: u64 (or Option<u64>), and kind: Creation/Deletion/DeletionWithDeposit?

I feel like the focus here can be on size / deposit as opposed to the op itself, e.g. given that the enum is called WriteOpSize and some variants have (u64), it seems an easy mistake for the reader to assume the u64 would be the size? Moreover, it seems like we could have deposit method alongside the write_len method and avoid straight up matches (I've commented on a few above, but those comments apply nonetheless)

igor-aptos replied:
as mentioned above comments cannot be addressed.

for here, I'll change to have name fields for enum options here - but otherwise type is generally what is matched on - and we have utility functions to get the fields you want

igor-aptos replied:
I'll use write_len as field name, as that is the name of the function - if you have preference for something else (i.e. value_size) let me know

igor-aptos replied:
I don't think it is cleaner to to have kind enum, and Options for other fields, that depend on the state of the enum.

I.e. kind=delete, size=Some would be invalid state, or kind=create, deposit=Some

with enum having fields that only refer to it, it is guaranteed to be internally consistent.

and so places that only care about one dimension (i.e. size), can get Option of it. cases that care about distinction will be able to match - and not need to deal with matching on create and then calling unwrap() on size, for example.

igor-aptos commented:
this is fully ready for review

igor-aptos commented on aptos-move/aptos-vm/src/move_vm_ext/respawned_session.rs:
```diff
@@ -342,9 +356,12 @@ impl<'r> TResourceGroupView for ExecutorViewWithChangeSet<'r> {
+    ) -> anyhow::Result<ResourceGroupSize> {
         // In respawned session, gas is irrelevant, so we return 0 (GroupSizeKind::None).
-        Ok(0)
+        Ok(ResourceGroupSize::zero_concrete())
```
I think it still applies

gelash commented on aptos-move/aptos-vm-types/src/output.rs:
```diff
@@ -135,13 +146,6 @@ impl VMOutput {
                 .is_empty(),
             "Delayed fields must be empty after materialization."
         );
-        assert!(
```
I guess there is no longer an easy way to check resource groups have been combined? can/do we check instead (here or in extend_resource_write_set, maybe in debug mode only) that all required abstract ops have been patched? That seems like a super useful check potentially (now we may even make it a PanicError it seems?)

gelash commented on aptos-move/aptos-vm-types/src/output.rs:
```diff
@@ -135,13 +146,6 @@ impl VMOutput {
                 .is_empty(),
             "Delayed fields must be empty after materialization."
```
can this be a panicerror now?

gelash commented on aptos-move/aptos-vm-types/src/resolver.rs:
```diff
@@ -283,3 +283,51 @@ pub trait StateValueMetadataResolver {
+    },
+}
+
+pub fn size_u32_as_uleb128(mut value: usize) -> usize {
```
Can we have a trivial unit test for this? (probably better than docs, as I couldn't immediately tell what it does)

gelash commented on aptos-move/aptos-vm-types/src/resolver.rs:
```diff
@@ -283,3 +283,51 @@ pub trait StateValueMetadataResolver {
 }
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum ResourceGroupSize {
```
Either here or at get, could you add a brief description for the Combined logic?

gelash commented on aptos-move/aptos-vm-types/src/storage.rs:
```diff
@@ -125,8 +125,8 @@ impl StoragePricingV2 {
 
-    fn write_op_size(&self, key: &StateKey, value: &[u8]) -> NumBytes {
-        let value_size = NumBytes::new(value.len() as u64);
+    fn write_op_size(&self, key: &StateKey, value_size: u64) -> NumBytes {
```
maybe value size? as WriteOpSize is now a different type

igor-aptos replied:
oh I addressed your comment, but then thought comment referred to the name of the function, not the argument :D

gelash reviewed (commented):
cont'd

gelash reviewed (commented):
penultimate comments from this pass

gelash commented on aptos-move/aptos-vm-types/src/change_set.rs:
```diff
@@ -219,104 +406,106 @@ impl VMChangeSet {
-            ));
-        }
-        Ok(self.into_storage_change_set_unchecked())
+    pub fn concrete_write_set_iter(&self) -> impl Iterator<Item = (&StateKey, Option<&WriteOp>)> {
```
I think I inspected all call-sites of this method and seems we can make it return &WriteOp instead of Option? Non-test uses are only in genesis, so it seems we might even be able to go further and extract and re-use the asserts for aggregator_v1 & delayed fields being empty with try_into_storage_change_set

gelash commented on aptos-move/aptos-vm-types/src/change_set.rs:
```diff
@@ -370,13 +554,35 @@ impl VMChangeSet {
-        );
+    ) -> Result<(), PanicError> {
+        for (k, new_write) in patched_resource_writes {
+            match self.resource_write_set.entry(k) {
```
completely optional re:style, but in the case when Vacant leads to error, might be worth to consider mut entry = ...get_mut(k).ok_or_else(code_invariant_error..)?;
and then the occupied(mut o/entry) logic - less indentation.

gelash reviewed (commented):
almost done with change_set changes, up to squash_additional_resource_writes. Since (for me and typically) that's the most error prone part, I'll do that separately next.
to echo the conversation there we should def try to simplify and test that in particular.

msmouse reviewed (commented):
Are we landing this any time soon?

msmouse reviewed (commented):
Are we landing this any time soon?

georgemitenkov commented:
Yes! But I need to check with others cause it might be easier to rebase this

igor-aptos commented:
this is needed for 1.9, and I would prefer to land it asap - as soon as it gets reviewed.

I've addressed all review comments (except from ones you just made)

gelash commented on aptos-move/aptos-vm/src/block_executor/mod.rs:
```diff
@@ -188,38 +205,46 @@ impl BlockExecutorTransactionOutput for AptosTransactionOutput {
         &self,
-    ) -> BTreeMap<
+    ) -> Vec<(
         <Self::Txn as BlockExecutableTransaction>::Key,
```
not sure but does it work to just have StateKey and WriteOp here? (and a few other places, just noticed that it seems not uniform)

igor-aptos replied:
yeah, this is just what IDE automatically fills

gelash commented on aptos-move/aptos-vm/src/move_vm_ext/respawned_session.rs:
```diff
@@ -329,8 +329,22 @@ impl<'r> TResourceView for ExecutorViewWithChangeSet<'r> {
+                    ..
+                }),
+            ) => Ok(write_op.as_state_value()),
+            // TODO[agg_v2](fix): Is this really unreachable, or is this called for metadata,
```
Looks like you implemented the dedicated metadata method? we should be able to get rid of the todo and keep unreachable in that case? (I'd add "why unreachable" to unreachable though)

gelash commented on aptos-move/aptos-vm/src/move_vm_ext/respawned_session.rs:
```diff
@@ -355,8 +398,16 @@ impl<'r> TResourceGroupView for ExecutorViewWithChangeSet<'r> {
+            .and_then(|write| match write {
+                AbstractResourceWriteOp::WriteResourceGroup(group_write) => Some(Ok(group_write)),
+                AbstractResourceWriteOp::ResourceGroupInPlaceDelayedFieldChange(_) => None,
+                _ => Some(Err(anyhow::anyhow!(
```
ditto _ =>, if we want to get rid of those as a general principle.

gelash commented on aptos-move/block-executor/src/executor.rs:
```diff
@@ -861,11 +869,15 @@ where
-                    Self::replace_ids_with_values(&value, layout.as_ref(), &latest_view),
-                );
+            for (key, layout) in reads_needing_delayed_field_exchange.into_iter() {
+                if let Ok(MVDataOutput::Versioned(_, ValueWithLayout::Exchanged(value, _))) =
```
could/should we (debug_)assert that layouts are equal?

igor-aptos replied:
added TODO for @vusirikala

igor-aptos replied:
yeah, they are in the wrong place, and @vusirikala  is moving the function, so leaving it for that PR

gelash reviewed (approved):
LGTM, thanks a lot!

igor-aptos commented on aptos-move/aptos-vm-types/src/tests/utils.rs:
```diff
@@ -39,15 +41,23 @@ macro_rules! as_state_key {
 macro_rules! as_bytes {
     ($v:ident) => {
-        serialize(&$v)
+        bcs::to_bytes(&$v).expect("unexpected serialization")
```
Word "error" is missing :)

georgemitenkov commented on aptos-move/aptos-vm-types/src/change_set.rs:
```diff
@@ -715,37 +843,167 @@ impl VMChangeSet {
+        for (key, additional_entry) in additional_write_set.into_iter() {
+            match write_set.entry(key.clone()) {
+                Occupied(mut entry) => {
+                    let (to_delete, to_overwite) = match (entry.get_mut(), &additional_entry) {
```
Let's do it separately

georgemitenkov commented on aptos-move/aptos-vm-types/src/change_set.rs:
```diff
@@ -715,37 +843,167 @@ impl VMChangeSet {
+                                materialized_size: additional_materialized_size,
+                            }),
+                        ) => {
+                            if layout != additional_layout {
```
yep

igor-aptos commented on aptos-move/aptos-vm-types/src/abstract_write_op.rs:
```diff
@@ -0,0 +1,219 @@
+    // No writes in the resource, except for delayed field changes.
+    InPlaceDelayedFieldChange(InPlaceDelayedFieldChangeOp),
+    // No writes in the resource group, except for delayed field changes.
+    ResourceGroupInPlaceDelayedFieldChange(ResourceGroupInPlaceDelayedFieldChangeOp),
```
we can rename later - if you have a better suggestion
